server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java

```diff
@@ -111,6 +113,27 @@
     public Mono<BlobId> save(BucketName bucketName, InputStream data) {
         Preconditions.checkNotNull(data);
 
+        return Mono.defer(() -> savingSrategySelection(bucketName, data));
+    }
+
+    private Mono<BlobId> savingSrategySelection(BucketName bucketName, InputStream data) {
+        InputStream bufferedData = new BufferedInputStream(data, BUFFERED_SIZE + 1);
+        try {
+            bufferedData.mark(0);
+            bufferedData.skip(BUFFERED_SIZE);
+            boolean isItABigStream = bufferedData.read() != -1;
+            bufferedData.reset();
```

Comment:
what if the IntputStream is not resetable?

Reply:
It's a `BufferedInputStream`, that's the point.

Reply:
Can't we just save `concat(buffer, remainingStream)`, which should be good as well, and would not require that hacky behaviour/knowledge?

Reply:
https://docs.oracle.com/javase/7/docs/api/java/io/BufferedInputStream.html

> A BufferedInputStream adds functionality to another input stream-namely, the ability to buffer the input and to support the mark and reset methods.

That's not a hack, that's the whole point of this class: providing read-ahead capability, that allow me to avoid to dealing with the recreation of an `InputStream`.